onnxruntime/core/optimizer/initializer.cc

```diff
@@ -11,49 +10,313 @@
+struct ToBFloat16;
 
-  std::vector<char> buffer(actual_tensor_data_length);
+template <>
+struct ToBFloat16<BFloat16> {
+  uint16_t operator()(const BFloat16& bf) const {
+    return bf.val;
+  }
+};
 
-  ORT_RETURN_IF_ERROR(Env::Default().ReadFileIntoBuffer(
-      (model_path.ParentPath() / external_data_relative_path).ToPathString().c_str(),
-      external_data->GetOffset(),
-      actual_tensor_data_length,
-      gsl::make_span(buffer)));
+template <>
+struct ToBFloat16<float> {
+  uint16_t operator()(float f) const {
+    return BFloat16(f).val;
+  }
+};
 
-  raw_data = std::move(buffer);
+template <>
+struct ToBFloat16<double> {
+  uint16_t operator()(double d) const {
+    return BFloat16(static_cast<float>(d)).val;
+  }
+};
 
-  return Status::OK();
+template <typename T>
+struct TensorToProtoBFloat16 {
+  void operator()(const Tensor& data, ONNX_NAMESPACE::TensorProto& proto) const {
+    ToBFloat16<T> to_bfloat16;
+    auto span = data.DataAsSpan<T>();
+    for (const auto& v : span) {
+      proto.add_int32_data(to_bfloat16(v));
+    }
+  }
+};
+
+inline void SetNameDims(const std::string& name,
+                        gsl::span<const int64_t> dims,
+                        ONNX_NAMESPACE::TensorProto_DataType dt,
+                        ONNX_NAMESPACE::TensorProto& tensor_proto) {
+  tensor_proto.set_name(name);
+  tensor_proto.set_data_type(dt);
+
+  for (auto d : dims) {
+    tensor_proto.add_dims(d);
+  }
 }
-}  // namespace onnxruntime
 
-#endif  // !(ORT_MINIMAL_BUILD)
+}  // namespace
+
+ONNX_NAMESPACE::TensorProto Initializer::ToFP16(const std::string& name) const {
+  ONNX_NAMESPACE::TensorProto tensor_proto;
+  SetNameDims(name, data_.Shape().GetDims(), ONNX_NAMESPACE::TensorProto_DataType_FLOAT16, tensor_proto);
+  utils::MLTypeCallDispatcher<MLFloat16, float, double> t_disp(data_.GetElementType());
+  t_disp.Invoke<TensorToProtoFP16>(data_, tensor_proto);
+  return tensor_proto;
+}
+
+ONNX_NAMESPACE::TensorProto Initializer::ToBFloat16(const std::string& name) const {
+  ONNX_NAMESPACE::TensorProto tensor_proto;
+  SetNameDims(name, data_.Shape().GetDims(), ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16, tensor_proto);
+  utils::MLTypeCallDispatcher<BFloat16, float, double> t_disp(data_.GetElementType());
+  t_disp.Invoke<TensorToProtoBFloat16>(data_, tensor_proto);
+  return tensor_proto;
+}
+
+namespace {
+
+// std::identity c++20
+template <typename T>
+struct ToNumeric {
+  using type = T;
+  constexpr const T& operator()(const T& v) const {
+    return v;
+  }
+};
+
+template <>
+struct ToNumeric<MLFloat16> {
+  using type = float;
+  float operator()(const MLFloat16& v) const {
+    return v.ToFloat();
+  }
+};
+
+template <>
+struct ToNumeric<BFloat16> {
+  using type = float;
+  float operator()(const BFloat16& v) const {
+    return v.ToFloat();
+  }
+};
+
+template <typename T, typename Op>
+struct OpElementWise {
+  void Invoke(Tensor& lhs, const Tensor& rhs) const {
+    Op op;
+    ToNumeric<T> to_numeric;
```

Comment:
Are these variables or just methods? Do they need to be declared?

Reply:
The are specialized structs. This works similar to std::hash